Source/SwiftLintFramework/Rules/Metrics/ScopeDepthRule.swift

```diff
@@ -0,0 +1,209 @@
+        }
+        return violations
+    }
+
+    private func determineStructureDepths(
+        _ structure: SwiftLintFramework.SourceKittenDictionary, depth: Int = 0
+    ) -> [(Int, SourceKittenDictionary)] {
+        var results = [(Int, SourceKittenDictionary)]()
+
+        results.append((depth, structure))
+
+        guard !structure.substructure.isEmpty else {
+            return results
+        }
+
+        // There's no point going deeper than we have to. If somthing is too
+        // deep, then obviously everything deeper will also need to be corrected.
+        guard depth <= configuration.errorDepth else {
+            return results
+        }
+
+        // Determine if we should increase the scope level or not
+        let shouldIncreaseScope: Bool = {
+            // If it was an expression it doesn't change the depth
+            if structure.expressionKind != nil {
+                return false
+            }
+
+            // If it was a declaration, it depends on the type of declaration
+            if let declarationKind: SwiftDeclarationKind = structure.declarationKind {
+                switch declarationKind {
+                case .`associatedtype`, .enumcase, .enumelement, .genericTypeParam,
+                     .module, .`typealias`, .varClass, .varGlobal, .varInstance,
+                     .varLocal, .varParameter, .varStatic:
+                  return false
+                case .`class`, .`enum`, .`extension`, .extensionClass, .extensionEnum,
+                     .extensionProtocol, .extensionStruct, .functionAccessorAddress,
+                     .functionAccessorDidset, .functionAccessorGetter, .functionAccessorModify,
+                     .functionAccessorMutableaddress, .functionAccessorRead, .functionAccessorSetter,
+                     .functionAccessorWillset, .functionConstructor, .functionDestructor,
+                     .functionFree, .functionMethodClass, .functionMethodInstance,
+                     .functionMethodStatic, .functionOperator, .functionOperatorInfix,
+                     .functionOperatorPostfix, .functionOperatorPrefix, .functionSubscript,
+                     .opaqueType, .precedenceGroup, .`protocol`, .`struct`:
+                  return true
+                }
+            }
+
+            // Statements usually increase scope/depth. For anything with a
+            // brace block though we just let the braces handle it (since they
+            // are separate from the statement).
+            if let statementKind = structure.statementKind {
+                switch statementKind {
+                case .brace, .`switch`, .`case`:
+                    return true
+                case .forEach, .`if`, .repeatWhile, .`guard`, .`while`, .`for`:
+                    return false
+                }
+            }
+
+            return false
+        }()
+
+        for substructure in structure.substructure {
+            let newDepth = shouldIncreaseScope ? depth + 1 : depth
```

Comment:
Can you return depth from the transform block of `SourceKittenDictionary`'s depth first search method?

I'm not sure if it supports early return. But would avoid implementing a similar traversal algorithm.

Reply:
I know basically nothing about SourceKitten, can you explain a little more about what you mean here?

Reply:
`SourceKittenDictionary` has     
`func traverseDepthFirst<T>(traverseBlock: (SourceKittenDictionary) -> [T]?) -> [T] {`

I only wonder if you traverse the root nodes in a file with `T` = `Depth`. and return if any `Depth` in `[Depth]` >= the configured limits. Only wondering if anyone is going get hung up on multiple DFS implementations against SourceKittens AST data structure.

Reply:
I wasn't aware of that, but it would be a little more complex to use anyway. As for returning early, we already do that where we can (but can't do it in all cases because we need _all_ violations, not just the first). Once we reach the error depth, we don't go any deeper.